### ISSUE TITLE
'gbk' codec can't decode bytes: illegal multibyte sequence

### ISSUE THREAD

**Issue by cpangam** [mattgwwalker/msg-extractor#103]
**Bug Metadata**
* Using latest supported version of extract_msg. (extract_msg-0.23.2-py2.py3-none-any.whl)
* Your python version: Python 2.7 
* How did you launch extract_msg?
  - I used the extract_msg package

**Describe the bug**
I have an app which extract the email files. If the email file contains `gbk` codec (Chinese characters), the Message(path to file) throws an exception.

**What code did you use or can we use to reproduce this error?**

```
from extract_msg import Message
try:
     msg = Message(vault_path)
except Exception as e:
     return action_result.set_status(APP_ERROR, "Failed to parse message. Error: {0}".format(str(e)))

```
**The error message from library call**
```
"message":"Failed to parse message. Error: 'gbk' codec can't decode bytes in position 34-35: illegal multibyte sequence"
```
Is there a message.msg file you want to share to help us reproduce this?
I am not allowed to share the file. The file has gbk codec (Chinese characters)

I did find the error string in the log file. There is no as such Trace back available.

**Comment by TheElementalOfCreation** [comment]
We specifically need the exact error thrown by our library to be able to ecen start to fix this. Remove tge encapsulation of the try-except block and post the exception that is thrown please, otherwise this issue will be closed.

**Comment by TheElementalOfCreation** [comment]
So that error is saying their is some problem parsing your data. As such, the codec "gbk" is unable to parse it.

It would be much easier to potentially help in coming to a solution if I had the full traceback to the location of the error. If the decoding issue is happening in something like the header, body, etc. I can give you specific instructions on how to check that the program has properly identified the codes being used. Otherwise I'll have to either add more code to the developer sections or give you some code that will help.

One of three things has happened here:
1. There is an error in Python's gbk codec.
2. There is an error in your data.
3. The codec identified by the program is incorrect.

Can you see if this problem occurs in python 3?

**Comment by cpangam** [comment]
Here is the traceback. I will check on the python 3 shortly.

```
File "/opt/folder/dependencies/extract_msg/message.py", line 95, in __init__#012    self.body#012  
File "/opt/folder/dependencies/extract_msg/message.py", line 460, in body#012    self._body = self._getStringStream('__substg1.0_1000')#012 
File "/opt/folder/dependencies/extract_msg/message.py", line 166, in _getStringStream#012    return None if tmp is None else
tmp.decode(self.stringEncoding)#012UnicodeDecodeError: 'gbk' codec can't decode bytes in position 34-35: illegal multibyte sequence
```

**Comment by pudo** [comment]
FWIW, I'm also working on encodings in my fork at the moment, this table mapping codepage identifiers to their names might be useful: 

https://github.com/alephdata/msglite/blob/master/msglite/encoding.py

Basically it can be resolved against the message properties (Not sure the handling there is correct yet):

https://github.com/alephdata/msglite/blob/master/msglite/message.py#L80

**Comment by TheElementalOfCreation** [comment]
@pudo so far it looks like we can literally just pass the codepage integer into the decode function to get the codec. Aside from this, which may not actually be an issue with this method, we have not had problems so far.

**Comment by TheElementalOfCreation** [comment]
@pudo I've tried it against other encodings without issues.

**Comment by cpangam** [comment]
@TheElementalOfCreation @pudo I tried with **ExtractMsg-0.3-py2-none-any.whl** module. I am facing the same issue.
Is there any alternative to resolve this **gbk** Unicode error?

**Comment by TheElementalOfCreation** [comment]
@cpangam angam

**Comment by TheElementalOfCreation** [comment]
@cpangam it looks like you are saying that you used version 0.3 for python 2. Simply download the code, and run setup.py with python 3 to get a python3 environment with the latest version (0.23.3 or something like that)

TLDR: you used one of the oldest versions in existence

**Comment by cpangam** [comment]
Our platform is based on python2.7 and my app is supposed to work on that platform. So I have to find the solution in python 2 only.
Our existing app code which available on our production has older version of wheel file. Recently, I have updated the module and did some minor changes in the code. Then we got issue from the customer related gbk codec.
I tried older version and newer version of app code on python2 and it is failing using both wheel.

**Comment by TheElementalOfCreation** [comment]
> Our platform is based on python2.7 and my app is supposed to work on that platform. So I have to find the solution in python 2 only.
> Our existing app code which available on our production has older version of wheel file. Recently, I have updated the module and did some minor changes in the code. Then we got issue from the customer related gbk codec.
> I tried older version and newer version of app code on python2 and it is failing using both wheel.

What I am trying to narrow down is whether the issue is specific to python 2 or not. If it is, then I'll know where to start looking to fix the issue

**Comment by cpangam** [comment]
Hey,
I have tried the same thing in python 3.8 and it is failing with the same error message. Here is the traceback.
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/extract_msg/message.py", line 95, in __init__
    self.body
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/extract_msg/message.py", line 460, in body
    self._body = self._getStringStream('__substg1.0_1000')
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/extract_msg/message.py", line 166, in _getStringStream
    return None if tmp is None else tmp.decode(self.stringEncoding)
UnicodeDecodeError: 'gbk' codec can't decode byte 0x97 in position 34: illegal multibyte sequence
```

**Comment by TheElementalOfCreation** [comment]
Thank you for that, I'll generate some test code to help try to identify the problem further, although it will take a bit of time for me to write.

**Comment by TheElementalOfCreation** [comment]
Okay, sorry it took me so long to get this to you, but here it is. First you'll need to open the msg file causing the error in outlook or a similar program and view the contents of the body. Then, compare the beginning of the body to the results of the code below. If they do not match, then the encoding of the msg file has somehow been misdetected, and we will need to fix that.
```python
import olefile

msg = olefile.OleFileIO(vault_path)
b = msg.exists('__substg1.0_1000001E')
with msg.openstream('__substg1.0_1000001E' if b else '__substg1.0_1000001F') as f:
    body = f.read()
msg.close()
print(body[:34].decode('gbk'))
```

Make sure this code is running in a console that supports unicode characters, otherwise it will just be a mess of escaped characters, a mess of gibberish, or may just throw an exception. Please tell me the results of the comparison between the body in Outlook and the portion of the body printed out by the program. Thank you.

**Comment by TheElementalOfCreation** [comment]
@cpangam Have you been able to run my test yet?

**Comment by cpangam** [comment]
@TheElementalOfCreation @mjiyani-crest and me did try this and it is not working. It is giving the same error.

**Comment by TheElementalOfCreation** [comment]
@cpangam you tried running the code snippet I provided and it gave the same error? If so, can you show me exactly what the error text was? Also, the code takes a substring of the text, so try lowering that number by one until the error stops appearing, but making sure not to go below at least 5.

**Comment by cpangam** [comment]
@TheElementalOfCreation We did try the snippet you provided. It is working as it is but as we remove that numeric value in the print statement, it fails to decode.
@mjiyani-crest is my team member and can provide more insight into this.

**Comment by TheElementalOfCreation** [comment]
@cpangam Could you open your outlook and check what encoding it is using for emails? Go to File > Options > Advanced and scroll down to "International Options" And tell me what it is set to for "Preferred encoding for outgoing messages"? I believe this may help me in determining what the encoding is that is being used for this email. Thank you.

**Comment by joanrodriguezr** [comment]
Hi all, 
I have similar errors, so I think I don't need to create a new issue. Those errors are:

'cp950' codec can't decode byte 0x96 in position 414: illegal multibyte sequence
'cp950' codec can't decode byte 0x96 in position 456: illegal multibyte sequence
'cp949' codec can't decode byte 0xa6 in position 1112: illegal multibyte sequence

Due to company policy I can't add there the .msg files but I you need me to repeat my tests against the msg files provoking those Errors I can try again adding any extra log in your library.

I'm using version 0.25.3 with Python v3.7.0:1bf9cc5093 and executing the library from a .py file.

Thanks in advance!

**Comment by TheElementalOfDestruction** [comment]
@joanrodriguezr First of all, it would be appreciated if you could move to the most recent version of msg extractor (0.26.2)? Secondly, please follow the instructions in the post immediately above your own.

**Comment by joanrodriguezr** [comment]
Ok, no problem I will update to 0.26.2 tomorrow and will let you know if the problem persists. (I thought there was no newer version than 0.25.3)

Meanwhile and related to the post above:

**"Preferred encoding for outgoing messages"** : Western European (ISO) 

This matches for 2 out of the 5 messages which I found the 3 errors, I'm waiting for someone to provide me the other 3 messages just to let you know which the Preferred encoding for outgoing messages is, if there's other different encoding I will let you know as well.

**Comment by joanrodriguezr** [comment]
I just have tested with extract_msg-0.26.3.dist-info and the errors remain the same for the 5 messages I tested:

'cp950' codec can't decode byte 0x96 in position 414: illegal multibyte sequence
'cp950' codec can't decode byte 0x96 in position 456: illegal multibyte sequence
'cp949' codec can't decode byte 0xa6 in position 1112: illegal multibyte sequence

Can I help testing anything else for you?

**Comment by TheElementalOfDestruction** [comment]
This issue should be fixed now. Please inform us if you continue to have errors.

**Comment by TheElementalOfDestruction** [comment]
We are assuming that this issue is fixed. If that is incorrect, please reopen and let us know.

**Comment by TheElementalOfDestruction** [comment]
So it looks like there may be more to this issue. I noticed that cp950 from Python cannot properly decode some of the files from Outlook, and it has to do with the Microsoft version being different from the one that Python uses. A previous version fixed one issue of encodings not properly being mapped, but if you are correctly having your text identified as CP950 then you are likely to run into issues still. The next coming version, 0.42.0, has a lot of bug fixes and feature additions, including an implementation for Microsoft's version of cp950. If anyone still has test files that were failing from that codec, you can test against the next-release branch and see if that fixes the issue. There is currently a known issue with RTFDE which is that it will use the Python version of the encoding still, but I'm seeing if I can manually patch the way it fetches that information from `extract-msg`.

Because of these encoding issues, I plan to look into other codecs to see if the Python implementation differs from the one Microsoft uses, as they will need a manual implementation as well. If the issue has to do with missing multibyte sequences, it shouldn't be too hard to implement a new encoding for them using the code I used for cp950.

I've created a new issue for these encoding mismatches, #373

(Also, I read through some of my responses during all of this and wow I sound so ignorant. Sorry about that, I've improved a lot and actually know how to listen when people try to explain an issue to me.)
